### PR TITLE
Simplify Fusion Definitions for Complex Functions

### PR DESCRIPTION
**Pull request by rdspring1**
Create NvFuser functions that are analogous with their eager mode counterparts.
e.g., softplus, dropout, softmax, batch_norm, layer_norm

1. Create ops directory to hold all fusion definitions
2. Use named variables for ops with multiple outputs
3. Import BERT and Scale+Bias+Relu benchmark tests

### CODE REVIEW

**Comment by rdspring1** [comment]
I finished prototyping a NvFuser API equivalent to the Aten eager mode functions. If we're happy with the design, then the PR is ready for review.

**Review by csarofeen — COMMENTED** [review]
Can we stick to using arith.cpp internally in the codegeneration unless all_ops is needed I think arith typically has everything we need for fusion/lowering/codegen.

File: torch/csrc/jit/codegen/cuda/ops/normalization.cpp [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+
+  auto c1 = mul(grad_x_hat, x_hat);
+  auto c2 = sum(c1, inner_reduction_axes);
+  auto bcast_c2 = broadcast(c2, inner_broadcast_mask);
+  auto c3 = mul(x_hat, bcast_c2);
+
+  auto inner = sub(sub(a, bcast_b), c3);
+  auto reciprocal_size = unaryOp(UnaryOpType::Reciprocal, num_features);
+
+  TensorView* dx = nullptr;
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/ops/normalization.cpp:185]: Can this actually be false?

**Review by csarofeen — COMMENTED** [review]
We need to default normalization to use welford. I just merged in welford translation during segmentatoin from Shiming. CC @shmsong

File: benchmarks/cpp/nvfuser/bert.cpp [hunk 2]
```diff
@@ -0,0 +1,917 @@
...
+  auto w = benchmark_state.range(0);
+  auto x = benchmark_state.range(1);
+  auto y = benchmark_state.range(2);
+  auto z = benchmark_state.range(3);
+
+  setupDivMaxSoftmaxDropoutForward(&fusion, dtype);
+
+  auto tvs = scheduler_utils::allTvs(&fusion);
+
+  std::vector<TensorView*> reduction_tvs;
```
Comment (csarofeen) [inline benchmarks/cpp/nvfuser/bert.cpp]: reduction and other tvs are no longer necessary for scheduler.

File: benchmarks/cpp/nvfuser/bert.cpp [hunk 3]
```diff
@@ -0,0 +1,917 @@
...
+  auto w = benchmark_state.range(0);
+  auto x = benchmark_state.range(1);
+  auto y = benchmark_state.range(2);
+  auto z = benchmark_state.range(3);
+
+  setupDivMaxSoftmaxDropoutBackward(&fusion, dtype);
+
+  auto tvs = scheduler_utils::allTvs(&fusion);
+
+  std::vector<TensorView*> reduction_tvs;
```
Comment (csarofeen) [inline benchmarks/cpp/nvfuser/bert.cpp]: reduction and other tvs are no longer necessary for scheduler.

File: benchmarks/cpp/nvfuser/bert.cpp [hunk 4]
```diff
@@ -0,0 +1,917 @@
...
+
+  auto x = benchmark_state.range(0);
+  auto y = benchmark_state.range(1);
+  auto z = benchmark_state.range(2);
+
+  setupBiasDropoutAddLayernormFwd(&fusion, dtype);
+
+  auto tvs = scheduler_utils::allTvs(&fusion);
+
+  std::vector<TensorView*> reduction_tvs;
```
Comment (csarofeen) [inline benchmarks/cpp/nvfuser/bert.cpp]: reduction and other tvs are no longer necessary for scheduler.

File: benchmarks/cpp/nvfuser/bert.cpp [hunk 5]
```diff
@@ -0,0 +1,917 @@
...
+
+  auto x = benchmark_state.range(0);
+  auto y = benchmark_state.range(1);
+  auto z = benchmark_state.range(2);
+
+  setupBiasDropoutAddLayernormBwd1(&fusion, dtype);
+
+  auto tvs = scheduler_utils::allTvs(&fusion);
+
+  std::vector<TensorView*> reduction_tvs;
```
Comment (csarofeen) [inline benchmarks/cpp/nvfuser/bert.cpp]: reduction and other tvs are no longer necessary for scheduler.

File: benchmarks/cpp/nvfuser/bert.cpp [hunk 6]
```diff
@@ -0,0 +1,917 @@
...
+
+  auto x = benchmark_state.range(0);
+  auto y = benchmark_state.range(1);
+  auto z = benchmark_state.range(2);
+
+  setupBiasDropoutAddLayernormBwd2(&fusion, dtype);
+
+  auto tvs = scheduler_utils::allTvs(&fusion);
+
+  std::vector<TensorView*> reduction_tvs;
```
Comment (csarofeen) [inline benchmarks/cpp/nvfuser/bert.cpp]: reduction and other tvs are no longer necessary for scheduler.

File: benchmarks/cpp/nvfuser/bert.cpp [hunk 7]
```diff
@@ -0,0 +1,917 @@
...
+
+  auto x = benchmark_state.range(0);
+  auto y = benchmark_state.range(1);
+  auto z = benchmark_state.range(2);
+
+  setupBiasDropoutAddLayernormBwd3(&fusion, dtype);
+
+  auto tvs = scheduler_utils::allTvs(&fusion);
+
+  std::vector<TensorView*> reduction_tvs;
```
Comment (csarofeen) [inline benchmarks/cpp/nvfuser/bert.cpp]: reduction and other tvs are no longer necessary for scheduler.

File: benchmarks/cpp/nvfuser/gelu_backward.cpp [hunk 8]
```diff
@@ -1,7 +1,7 @@
 
 // Based on NVFuserTest.FusionBiasGeluBwd_CUDA
 
-#include <torch/csrc/jit/codegen/cuda/arith.h>
+#include <torch/csrc/jit/codegen/cuda/ops/all_ops.h>
```
Comment (csarofeen) [inline benchmarks/cpp/nvfuser/gelu_backward.cpp]: should we put a todo to add a backward gelu op in composite?

File: benchmarks/cpp/nvfuser/lstm_cell.cpp [hunk 9]
```diff
@@ -1,4 +1,4 @@
-#include <torch/csrc/jit/codegen/cuda/arith.h>
+#include <torch/csrc/jit/codegen/cuda/ops/all_ops.h>
```
Comment (csarofeen) [inline benchmarks/cpp/nvfuser/lstm_cell.cpp]: Should we add a TODO to add LSTM cell ops in composite?

**Review by csarofeen — COMMENTED** [review]
Really nice PR. Minor comments. I'm not approving yet just because I want @jjsjann123 to review.

File: benchmarks/cpp/nvfuser/lstm_cell.cpp (hunk 9 above)
Comment (rdspring1) [reply benchmarks/cpp/nvfuser/lstm_cell.cpp]: What is the signature for this LSTM function?
`cy, hy = lstm(x, cx)`

File: torch/csrc/jit/codegen/cuda/predicate_compute.h [hunk 10]
```diff
@@ -1,9 +1,9 @@
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/predicate_compute.h]: nitpick: empty new line.

File: torch/csrc/jit/codegen/cuda/ops/normalization.cpp [hunk 11]
```diff
@@ -0,0 +1,412 @@
...
+    TensorView* dy,
+    TensorView* weight,
+    TensorView* running_mean,
+    TensorView* running_var,
+    TensorView* save_mean,
+    TensorView* save_invstd,
+    const bool kTraining,
+    Val* eps,
+    const std::vector<bool>& output_mask) {
+  // TODO: merge this loop below.
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/ops/normalization.cpp]: nitpick: can we put this down inside the other if block?

File: torch/csrc/jit/codegen/cuda/ops/normalization.cpp (hunk 1 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/ops/normalization.cpp:185]: I think so. If you have a trivial example with `input(requires_grad=false)->layer_norm`

**Review by jjsjann123 — COMMENTED** [review]
LGTM

**Review by jjsjann123 — APPROVED** [review]
stamping for merge
